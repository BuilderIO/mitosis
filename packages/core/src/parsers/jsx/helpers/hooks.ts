import * as babel from '@babel/core';

const { types } = babel;

export const getHook = (node: babel.Node) => {
  const item = node;
  if (types.isExpressionStatement(item)) {
    const expression = item.expression;
    if (types.isCallExpression(expression)) {
      if (types.isIdentifier(expression.callee)) {
        return expression;
      }
    }
  }
  return null;
};
