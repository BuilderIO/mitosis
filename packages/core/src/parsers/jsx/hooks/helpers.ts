import * as babel from '@babel/core';
import generate from '@babel/generator';

const { types } = babel;

export const getHook = (node: babel.Node) => {
  const item = node;
  if (types.isExpressionStatement(item)) {
    const expression = item.expression;
    if (types.isCallExpression(expression)) {
      if (types.isIdentifier(expression.callee)) {
        return expression;
      }
    }
  }
  return null;
};

export const processHookCode = (
  firstArg: babel.types.ArrowFunctionExpression | babel.types.FunctionExpression,
) =>
  generate(firstArg.body)
    .code.trim()
    // Remove arbitrary block wrapping if any
    // AKA
    //  { console.log('hi') } -> console.log('hi')
    .replace(/^{/, '')
    .replace(/}$/, '')
    .trim();
