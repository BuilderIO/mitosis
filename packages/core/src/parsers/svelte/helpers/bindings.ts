import type { MitosisNode } from '../../../types/mitosis-node';
import type { SveltosisComponent } from '../types';

function replaceGroupWithChecked(node: MitosisNode, isArray = false) {
  if (node.bindings.group?.code?.length) {
    const bindingValue = node.bindings.value?.code;
    const propertyValue = node.properties.value;
    const groupBinding = node.bindings.group.code;

    let code = '';

    if (isArray) {
      code = bindingValue
        ? `${groupBinding}.includes(${bindingValue})`
        : `${groupBinding}.includes('${propertyValue}')`;
    } else {
      code = bindingValue
        ? `${groupBinding} === ${bindingValue}`
        : `${groupBinding} === '${propertyValue}'`;
    }

    node.bindings['checked'] = {
      code,
    };
    delete node.bindings.group;
  }
}

/* post-processes bindings
  bind:group https://svelte.dev/docs#template-syntax-element-directives-bind-group
  bind:property https://svelte.dev/docs#template-syntax-component-directives-bind-this
  bind:this https://svelte.dev/docs#template-syntax-component-directives-bind-this

  - replaces group binding with checked for checkboxes and radios (supported inputs for bind:group)
  - adds onChange for bind:group and bind:property (event.target.value OR event.target.checked)
*/
export function processBindings(json: SveltosisComponent, node: MitosisNode) {
  let name;
  let target = 'event.target.value';
  let binding = '';
  let isArray = false;

  if (Object.prototype.hasOwnProperty.call(node.bindings, 'group')) {
    name = 'group';
    binding = node.bindings.group?.code ?? '';

    if (binding.startsWith('state.')) {
      const stateObject = json.state[binding.replace(/^state\./, '')];
      isArray = Array.isArray(stateObject?.code);
    }

    replaceGroupWithChecked(node, isArray);

    if (node.properties.type === 'checkbox' && !node.properties.value) {
      target = 'event.target.checked';
    }
  } else if (Object.prototype.hasOwnProperty.call(node.bindings, 'this')) {
    name = 'ref';
    binding = node.bindings.this?.code ?? '';
  } else if (
    Object.prototype.hasOwnProperty.call(node.bindings, 'onChange') &&
    node.properties.type === 'checkbox'
  ) {
    target = 'event.target.checked';
    binding = node.bindings.onChange?.code.split('=')[0] ?? '';
  }

  let onChangeCode = `${binding} = ${target}`;

  // If the binding is an array, we should push / splice rather than assigning
  if (isArray) {
    onChangeCode = `event.target.checked ? ${binding}.push(${target}) : ${binding}.splice(${binding}.indexOf(${
      node.properties.value ? `'${node.properties.value}'` : node.bindings.value?.code
    }), 1)`;
  }

  if (name !== 'ref' && binding) {
    node.bindings['onChange'] = {
      code: onChangeCode,
      arguments: ['event'],
    };
  }
}
