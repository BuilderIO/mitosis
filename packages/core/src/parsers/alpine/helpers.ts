import { createSingleBinding } from '../../helpers/bindings';
import { MitosisNode } from '../../types/mitosis-node';

/**
 * Parses an Alpine.js directive into a Mitosis binding
 */
export function parseDirective(directive: string, value: string): MitosisNode['bindings'] {
  const bindings: MitosisNode['bindings'] = {};

  switch (directive) {
    case 'x-text':
      bindings._text = createSingleBinding({ code: value });
      break;
    case 'x-data':
      // x-data will be handled separately as it defines the component's state
      break;
    default:
      // For other directives, create a binding with the directive name as the key
      bindings[directive] = createSingleBinding({ code: value });
  }

  return bindings;
}

/**
 * Parses an Alpine.js data object into a Mitosis state object
 */
export function parseDataObject(dataString: string): Record<string, any> {
  try {
    // Remove the outer curly braces and parse the JSON
    const jsonString = dataString.trim().replace(/^{|}$/g, '');
    return JSON.parse(`{${jsonString}}`);
  } catch (e) {
    // If parsing fails, return an empty object
    return {};
  }
} 