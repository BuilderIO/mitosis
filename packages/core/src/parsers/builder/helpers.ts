import { GETTER } from '../../helpers/patterns';
import { MitosisComponent, StateValue, StateValueType } from '../../types/mitosis-component';
import { JSONObject } from '../../types/json';
import { mapValues } from 'lodash';

const __DO_NOT_USE_FUNCTION_LITERAL_PREFIX = `@builder.io/mitosis/function:`;
const __DO_NOT_USE_METHOD_LITERAL_PREFIX = `@builder.io/mitosis/method:`;

/**
 * Maps the Builder State format to the Mitosis State format.
 */
const mapJsonToStateValue = (value: any): StateValue => {
  if (typeof value === 'string') {
    if (value.startsWith(__DO_NOT_USE_FUNCTION_LITERAL_PREFIX)) {
      return { type: 'function', code: value.replace(__DO_NOT_USE_FUNCTION_LITERAL_PREFIX, '') };
    } else if (value.startsWith(__DO_NOT_USE_METHOD_LITERAL_PREFIX)) {
      const strippedValue = value.replace(__DO_NOT_USE_METHOD_LITERAL_PREFIX, '');
      const isGet = Boolean(strippedValue.match(GETTER));
      const type: StateValueType = isGet ? 'getter' : 'method';
      return { type, code: strippedValue };
    }
  }
  return { type: 'property', code: JSON.stringify(value) };
};

export const mapBuilderContentStateToMitosisState = (
  value: JSONObject,
): MitosisComponent['state'] => mapValues(value, mapJsonToStateValue);
