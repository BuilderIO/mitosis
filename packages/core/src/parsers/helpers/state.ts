import { mapValues } from 'lodash';
import { GETTER } from '../../helpers/patterns';
import {
  __DO_NOT_USE_METHOD_LITERAL_PREFIX,
  __DO_NOT_USE_FUNCTION_LITERAL_PREFIX,
} from '../constants/outdated-prefixes';
import { JSONObject } from '../../types/json';
import { MitosisComponent, StateValue, StateValueType } from '../../types/mitosis-component';

/**
 * Sets StateTypeValue based on the string prefixes we've set previously.
 *
 * This is a temporary workaround until we eliminate the prefixes and make this StateValueType the
 * source of truth.
 */
const mapJsonToStateValue = (value: any): StateValue => {
  if (typeof value === 'string') {
    if (value.startsWith(__DO_NOT_USE_FUNCTION_LITERAL_PREFIX)) {
      return { type: 'function', code: value.replace(__DO_NOT_USE_FUNCTION_LITERAL_PREFIX, '') };
    } else if (value.startsWith(__DO_NOT_USE_METHOD_LITERAL_PREFIX)) {
      const strippedValue = value.replace(__DO_NOT_USE_METHOD_LITERAL_PREFIX, '');
      const isGet = Boolean(strippedValue.match(GETTER));
      const type: StateValueType = isGet ? 'getter' : 'method';
      return { type, code: strippedValue };
    }
  }
  return { type: 'property', code: value };
};

export const mapJsonObjectToStateValue = (value: JSONObject): MitosisComponent['state'] =>
  mapValues(value, mapJsonToStateValue);
