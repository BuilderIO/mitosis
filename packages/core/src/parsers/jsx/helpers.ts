import * as babel from '@babel/core';
import generate from '@babel/generator';
import { tryParseJson } from '../../helpers/json';
import { objectHasKey } from '../../helpers/typescript';

const { types } = babel;

export const uncapitalize = (str: string) => {
  if (!str) {
    return str;
  }

  return str[0].toLowerCase() + str.slice(1);
};

export const parseCode = (node: babel.types.Node) => {
  return generate(node).code;
};

export const parseCodeJson = (node: babel.types.Node) => {
  const code = parseCode(node);
  return tryParseJson(code);
};

export const isImportOrDefaultExport = (node: babel.Node) =>
  types.isExportDefaultDeclaration(node) || types.isImportDeclaration(node);

export const HTML_ATTR_FROM_JSX = {
  htmlFor: 'for',
};

export const transformAttributeName = (name: string) => {
  if (objectHasKey(HTML_ATTR_FROM_JSX, name)) return HTML_ATTR_FROM_JSX[name];
  return name;
};
