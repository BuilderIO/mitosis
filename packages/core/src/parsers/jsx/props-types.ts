import { Project, PropertySignature } from 'ts-morph';
import { getPropsSymbol } from '../../helpers/typescript-project';

export const findOptionalProps = (args: { project: Project; filePath: string }) => {
  const ast = args.project.getSourceFileOrThrow(args.filePath);

  if (ast === undefined) {
    throw new Error('Could not find AST. Please provide either `code` or `filePath` configs.');
  }

  const propsSymbol = getPropsSymbol(ast);

  if (!propsSymbol) return [];

  return propsSymbol
    .getDeclarations()[0]
    .getType()
    .getProperties()
    .map((p) => p.getDeclarations()[0])
    .filter(
      (k): k is PropertySignature =>
        PropertySignature.isPropertySignature(k) && k.hasQuestionToken(),
    )
    .map((k) => k.getName());
};
