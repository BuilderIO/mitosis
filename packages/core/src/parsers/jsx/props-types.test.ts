import { createTypescriptProject } from '../../helpers/typescript-project';
import { findOptionalProps } from './props-types';

describe(findOptionalProps.name, () => {
  /**
   * We can piggyback on the `core` project's TS config, since we are allowed to reference `@builder.io/mitosis`
   * recursively inside of itself.
   * This avoids the need to create a mock TS project just for testing.
   */
  const tsProject = createTypescriptProject(__dirname + '/../../../tsconfig.json');

  test('types', () => {
    const code = `
    type Kaboom = {
      foo?: string
    }
    
    type Props = Kaboom & {
      normal: string;
      bar: boolean
      id?: number;
    }
    
    export default function InlinedStyles(props: Props) {
      return props;
    }
    `;
    tsProject.project.createSourceFile('src/testing.tsx', code, { overwrite: true });

    const result = findOptionalProps({ filePath: 'src/testing.tsx', ...tsProject });
    expect(result).toMatchSnapshot();
  });

  test('interfaces', () => {
    const code = `
    type Kaboom = {
      foo?: string
    }
    
    interface Props extends Kaboom {
      styles: string;
      id?: string;
    }
    
    export default function InlinedStyles(props: Props) {
      return props;
    }
    `;
    tsProject.project.createSourceFile('src/testing.tsx', code, { overwrite: true });

    const result = findOptionalProps({ filePath: 'src/testing.tsx', ...tsProject });
    expect(result).toMatchSnapshot();
  });

  test('type extending interface', () => {
    const code = `
    interface Kaboom {
      foo?: string
    }
    
    type Props = Kaboom & {
      styles: string;
      id?: string;
    }
    
    export default function InlinedStyles(props: Props) {
      return props;
    }
    `;
    tsProject.project.createSourceFile('src/testing.tsx', code, { overwrite: true });

    const result = findOptionalProps({ filePath: 'src/testing.tsx', ...tsProject });
    expect(result).toMatchSnapshot();
  });
});
