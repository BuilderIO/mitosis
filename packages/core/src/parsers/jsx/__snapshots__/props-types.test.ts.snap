// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findOptionalProps > interfaces 1`] = `
[
  "id",
  "foo",
]
`;

exports[`findOptionalProps > type extending interface 1`] = `
[
  "foo",
  "id",
]
`;

exports[`findOptionalProps > types 1`] = `
[
  "foo",
  "id",
]
`;
